### PR TITLE
check externalNameService more time

### LINKED ISSUE #98104: [Flaky Test] [sig-network] Services should be able to change the type from ClusterIP to ExternalName
**Which jobs are flaking**:
- `ci-kubernetes-e2e-ubuntu-gce-containerd`
- `ci-kubernetes-e2e-gci-gce`

**Which test(s) are flaking**:
`[sig-network] Services should be able to change the type from ClusterIP to ExternalName`

**Testgrid link**:
- https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-ubuntu-master-containerd
- https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-cos-master-default

**Reason for failure**:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:640
Jan 15 09:43:10.121: Unexpected error:
    <*errors.errorString | 0xc002cf2840>: {
        s: "ExternalName service \"services-5563/execpod4sxhv\" must resolve to IP",
    }
    ExternalName service "services-5563/execpod4sxhv" must resolve to IP
occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:1773
```

**Anything else we need to know**:
Having some trouble with finding this in Triage, but here are two example failures:

- https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1350010433976668160
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-ubuntu-gce-containerd/1349741395811766272

/sig network

### PR DESCRIPTION
**Pull request by JornShen**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

add more check to externalNameService 

the failed test error as fellow:
```
Jan 15 09:43:03.808: INFO: Creating new exec pod
Jan 15 09:43:08.368: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://34.105.6.162 --kubeconfig=/workspace/.kube/config --namespace=services-5563 exec execpod4sxhv -- /bin/sh -x -c nslookup clusterip-service.services-5563.svc.cluster.local'
Jan 15 09:43:10.121: INFO: rc: 1
Jan 15 09:43:10.121: FAIL: Unexpected error:
    <*errors.errorString | 0xc002cf2840>: {
        s: "ExternalName service \"services-5563/execpod4sxhv\" must resolve to IP",
    }
    ExternalName service "services-5563/execpod4sxhv" must resolve to IP
occurred

Full Stack Trace
k8s.io/kubernetes/test/e2e/network.glob..func24.17()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:1773 +0x465
k8s.io/kubernetes/test/e2e.RunE2ETests(0xc0038b8480)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:130 +0x36c
k8s.io/kubernetes/test/e2e.TestE2E(0xc0038b8480)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:144 +0x2b
testing.tRunner(0xc0038b8480, 0x4fe76e8)
	/usr/local/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x2b3
```
It failed in checking `externalName` type service effective.
And I found the case only check it once.

https://github.com/kubernetes/kubernetes/blob/9f2f1b8c460b4b5ec8bfcd34490fb259ebad65b0/test/e2e/framework/service/jig.go#L991

I think check it once is not enough. The work item is DNS side. But we can control the order of dns work completed and
our check. So add more check times will be more reasonable.

**Which issue(s) this PR fixes:**

Fix https://github.com/kubernetes/kubernetes/issues/98104

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/network/service.go [hunk 1]
```diff
@@ -1436,7 +1444,12 @@ var _ = SIGDescribe("Services", func() {
 			framework.Failf("Spec.ClusterIP was not cleared")
 		}
 		execPod := e2epod.CreateExecPodOrFail(cs, ns, "execpod", nil)
-		err = jig.CheckServiceReachability(externalNameService, execPod)
+		err = wait.PollImmediate(framework.Poll, CheckExternalNameServiceTimeOut, func() (done bool, err error) {
```
Comment (aojea) [inline test/e2e/network/service.go:1447]: I think that is better to add the `wait.PollImmediate`  here, so the method is reliable and people don´t have to add the active loops on top
https://github.com/kubernetes/kubernetes/blob/252887e39f905389156d2bc9c5932688857588e4/test/e2e/framework/service/jig.go#L986-L997

Comment (JornShen) [reply test/e2e/network/service.go:1447]: right :), since `checkExternalServiceReachability` method only used by cases in service.go.

**Comment by JornShen** [comment]
updated. PTAL :)
@aojea

File: test/e2e/framework/service/jig.go [hunk 2]
```diff
@@ -988,13 +988,15 @@ func (j *TestJig) checkExternalServiceReachability(svc *v1.Service, pod *v1.Pod)
...
 	// Service must resolve to IP
 	cmd := fmt.Sprintf("nslookup %s", svcName)
-	_, stderr, err := framework.RunHostCmdWithFullOutput(pod.Namespace, pod.Name, cmd)
-	// NOTE(claudiub): nslookup may return 0 on Windows, even though the DNS name was not found. In this case,
-	// we can check stderr for the error.
-	if err != nil || (framework.NodeOSDistroIs("windows") && strings.Contains(stderr, fmt.Sprintf("can't find %s", svcName))) {
-		return fmt.Errorf("ExternalName service %q must resolve to IP", pod.Namespace+"/"+pod.Name)
-	}
-	return nil
+	return wait.PollImmediate(framework.Poll, CheckExternalNameServiceTimeOut, func() (done bool, err error) {
```
Comment (aojea) [inline test/e2e/framework/service/jig.go:991]: Just use the existing timeout ServiceReachabilityShortPollTimeout for consistency ...

Comment (JornShen) [reply test/e2e/framework/service/jig.go]: ohhh, ops, my problem 😂 
see updated.

File: test/e2e/framework/service/jig.go [hunk 3]
```diff
@@ -988,13 +988,15 @@ func (j *TestJig) checkExternalServiceReachability(svc *v1.Service, pod *v1.Pod)
...
-	if err != nil || (framework.NodeOSDistroIs("windows") && strings.Contains(stderr, fmt.Sprintf("can't find %s", svcName))) {
-		return fmt.Errorf("ExternalName service %q must resolve to IP", pod.Namespace+"/"+pod.Name)
-	}
-	return nil
+	return wait.PollImmediate(framework.Poll, ServiceReachabilityShortPollTimeout, func() (done bool, err error) {
+		_, stderr, err := framework.RunHostCmdWithFullOutput(pod.Namespace, pod.Name, cmd)
+		// NOTE(claudiub): nslookup may return 0 on Windows, even though the DNS name was not found. In this case,
+		// we can check stderr for the error.
+		if err != nil || (framework.NodeOSDistroIs("windows") && strings.Contains(stderr, fmt.Sprintf("can't find %s", svcName))) {
+			return false, fmt.Errorf("ExternalName service %q must resolve to IP", pod.Namespace+"/"+pod.Name)
```
Comment (aojea) [inline test/e2e/framework/service/jig.go:996]: sorry for not catching this before, but this loops seems that will only run once, or it succeeds or it fails ... I think that we need this, (check if the commands are correct please) but we should not return an error or we´ll never retry
```suggestion
                        framework.Logf("ExternalName service %q failed to resolve to IP", pod.Namespace+"/"+pod.Name)
			return false, nil
```

Comment (JornShen) [reply test/e2e/framework/service/jig.go:996]: You are right.We should keep retry if it failed.

**Review by MrHohn — APPROVED** [review]
/approve

**Comment by aojea** [comment]
/retest
storage flake

**Pull request merged**